mcm/json_layer/validation.py

```diff
@@ -0,0 +1,482 @@
+        except Exception:
+            pass
+
+        return target_output_events
+
+    def _get_test_min_output_events(self) -> int:
+        """
+        Get the number of minimum output events that the validation job
+        should produce to be considered valid.
+        """
+        min_output_events = 10
+        try:
+            min_output_events = int(settings.get_value("validation_min_output_events"))
+        except Exception:
+            pass
+
+        return min_output_events
+
+    def _test_on_output_events(self) -> bool:
+        """Determine whether to run the validation based on output events."""
+        test_on_output_events = True
+        try:
+            test_on_output_events = bool(settings.get_value("validation_test_on_output"))
+        except Exception:
+            pass
+
+        return test_on_output_events
+
+    def _get_max_number_validations(self) -> int:
+        """Get the number of maximum validation test jobs to run for a request."""
+        number_of_validation_tests = 4
+        try:
+            number_of_validation_tests = int(settings.get_value("validation_max_test_jobs"))
+        except Exception:
+            pass
+
+        return number_of_validation_tests
+
+    def _get_minimum_validation_runtime(self) -> int:
+        """Get the minimum validation runtime for a validation job in seconds."""
+        minimum_validation_runtime = 10 * 60 # 10 minutes
+        try:
+            minimum_validation_runtime = int(settings.get_value("validation_min_runtime"))
+        except Exception:
+            pass
+
+        return minimum_validation_runtime
+
+    def _get_thread_configurations(self, request) -> dict[str, dict | None]:
+        """
+        Get the number of output, input events, and other metadata
+        for each thread configuration to consider for sending jobs.
+
+        Args:
+            request (json_layer.request.request): Request to which validation
+                events will be counted.
+
+        Returns:
+            The input and output events for each thread to submit
+                validation jobs.
+        """
+        threads_configuration = {}
+        # Thread configuration to test
+        threads_for_test = [1, 2, 4, 8, 16, 32]
+        # Target output events expected for the validation
+        output_events = self._get_test_output_events()
+        # Minimum output events the validation must produce
+        min_output_events = self._get_test_min_output_events()
+        # Max number of seconds that validation can run for
+        max_runtime = self.get_validation_max_runtime(request=request)
+        # Minimum validation runtime
+        min_runtime = self._get_minimum_validation_runtime()
+        # Maximum number of validation jobs to run for a request
+        max_number_validations = max(self._get_max_number_validations(), 1)
+        # Event efficiency (a.k.a `filter_efficiency` in McM internals)
+        event_efficiency = request.get_efficiency()
```

Comment:
so as to cover enough of the low efficiency phase space, the new expression should actually be
```
eff = request.get_efficiency()
eff_err = request.get_efficiency_error()
event_efficiency = eff - 2*eff_err
if event_efficiency < 0:
   event_efficiency = eff - eff_err
if event_efficiency < 0:
   event_efficiency = efficiency/ 2.
```